### PR TITLE
[FIRRTL] enumcreate for creating enum values

### PR DESCRIPTION
**Pull request by darthscsi**
Add an operation to create enum values.  Extracted from here: https://github.com/llvm/circt/commit/d36dc6615a7230966170e1a253b506eaeda78324

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 1]
```diff
@@ -195,6 +195,40 @@ def VectorCreateOp : FIRRTLOp<"vectorcreate"> {
...
 }
 
+def FEnumCreateOp : FIRRTLOp<"enumcreate"> {
+  let summary = "Produce a enum value";
+  let description = [{
+    Create an enum from tag and value.
+    ```
+      %result = firrtl.enumcreate field-name(%input) : !firrtl.enum<None: uint<0>, Some: uint<8>>
+
+    ```
```
Comment (youngar) [inline include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:205]:
```suggestion
    ```mlir
      %result = firrtl.enumcreate field-name(%input) : !firrtl.enum<None: uint<0>, Some: uint<8>>
    ```
```

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 2]
```diff
@@ -195,6 +195,40 @@ def VectorCreateOp : FIRRTLOp<"vectorcreate"> {
...
+      return build($_builder, $_state, resultType, (uint32_t)*fieldIndex, input);
+    }]>
+  ];
+
+  let extraClassDeclaration = [{
+    /// Return the name of the accessed field.
+    StringRef getFieldName() {
+      return getResult().getType().cast<FEnumType>()
+        .getElementName(getFieldIndex());
+    }
```
Comment (youngar) [inline include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:228]: Might not need to `cast<FEnumType>` here

**Review by youngar — APPROVED** [review]

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td (hunk 2 above)
Comment (darthscsi) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:228]: and the custom builder was buggy.

File: test/Dialect/FIRRTL/errors.mlir [hunk 3]
```diff
@@ -1175,3 +1175,28 @@ firrtl.circuit "ForceableTypeMismatch" {
...
+  // expected-error @below {{expects different type than prior uses}}
+  %some = firrtl.enumcreate Some(%in) : !firrtl.enum<None: uint<0>, Some: uint<8>>
+}
+
+// -----
+
+firrtl.circuit "EnumCreateNoCase" {
+firrtl.module @EnumCreateNoCase(in %in : !firrtl.enum<None: uint<0>, Some: uint<8>>) {
+  // expected-error @below {{unknown field SomeOther in enum type}}
+  %some = firrtl.subtag %in[SomeOther] : !firrtl.enum<None: uint<0>, Some: uint<8>>
```
Comment (youngar) [inline test/Dialect/FIRRTL/errors.mlir:1201]: This test looks like its for the wrong operation

Comment (darthscsi) [reply test/Dialect/FIRRTL/errors.mlir:1201]: right op, wrong module name

**Pull request merged**